hawkular/hawkular-apm
# HWKAPM-495 | Test framework for checking instrumentation rules

Test scenarios are provided in a separate file structure which contains test scenario definition, scripts and possibly docker-compose.

Test scenario executes test environment (docker image, docker-compose). When executing docker image test framework it first instruments (only when type APM). When executing docker-compose users have to specify Hawkular APM specific properties (env variables, network). Example configuration files are provided

Test scenario:
* name
* environment definition (docker image)
* set of test cases

Test environment:
* image: docker image name,
* docker-compose: docker-compose name inside test directory
* initWaitSeconds: wait in ms after starting environment
* type: enum APM or ZIPKIN. Then type is APM jsonPath verifies captures traces objects, when ZIPKIN captured spans are verified.

Test case:
* description: string description
* action: (bash, python..) for invoking business transactions in the targeted app. The script is executed inside test environment.
* afterActionWaitSeconds: wait in ms after invoking action script
* verify: currently a set of JSON path expression, could be extended to use a shell script.
* skip: whether to skip or not

Life cycle of the test case:
1. create network (must for docker-compose, optional for single docker image)
2. start apm server
3. start environment (which includes start of the app)
4. wait for app start
5. run action in the app which is executed inside docker container.
6. wait some time that the action is processed and some data propagated to apm server
7. verify results.
8. shut down test environment and apm server.
